File: src/services/telemetry.ts
import { invoke } from '@tauri-apps/api/core';

export interface TelemetryEvent {
  name: string;
  properties?: Record<string, unknown>;
  timestamp: number;
  sessionId: string;
  version: string;
  platform: string;
}

export interface TelemetryMetrics {
  filesOpened: number;
  filesSaved: number;
  charactersTyped: number;
  commandsExecuted: number;
  terminalsCreated: number;
  terminalCommandsRun: number;
  gitOperations: number;
  gitCommits: number;
  gitPushes: number;
  gitPulls: number;
  extensionsInstalled: number;
  extensionsActivated: number;
  aiQueries: number;
  aiCompletionsAccepted: number;
  aiCompletionsRejected: number;
  startupTime: number;
  memoryUsage: number;
  cpuUsage: number;
  errors: number;
  crashes: number;
}

class TelemetryService {
  private sessionId: string;
  private version: string;
  private platform: string;
  private events: TelemetryEvent[] = [];
  private metrics: TelemetryMetrics = {
    filesOpened: 0,
    filesSaved: 0,
    charactersTyped: 0,
    commandsExecuted: 0,
    terminalsCreated: 0,
    terminalCommandsRun: 0,
    gitOperations: 0,
    gitCommits: 0,
    gitPushes: 0,
    gitPulls: 0,
    extensionsInstalled: 0,
    extensionsActivated: 0,
    aiQueries: 0,
    aiCompletionsAccepted: 0,
    aiCompletionsRejected: 0,
    startupTime: 0,
    memoryUsage: 0,
    cpuUsage: 0,
    errors: 0,
    crashes: 0,
  };
  private flushInterval: number | null = null;
  private enabled: boolean = false;

  constructor() {
    this.sessionId = this.generateSessionId();
    this.version = '0.1.0';
    this.platform = navigator.platform;
  }

  private generateSessionId(): string {
    return `session_${Date.now()}_${Math.random().toString(36).substring(2, 15)}`;
  }

  async initialize(enabled: boolean = false): Promise<void> {
    this.enabled = enabled;
    if (enabled) {
      this.startFlushInterval();
      this.trackEvent('session_start', { version: this.version });
    }
  }

  private startFlushInterval(): void {
    this.flushInterval = window.setInterval(() => {
      this.flush();
    }, 60000);
  }

  stop(): void {
    if (this.flushInterval) {
      clearInterval(this.flushInterval);
      this.flushInterval = null;
    }
    if (this.enabled) {
      this.trackEvent('session_end', { duration: Date.now() });
      this.flush();
    }
  }

  trackEvent(name: string, properties?: Record<string, unknown>): void {
    if (!this.enabled) return;
    this.events.push({
      name,
      properties,
      timestamp: Date.now(),
      sessionId: this.sessionId,
      version: this.version,
      platform: this.platform,
    });
  }

  incrementMetric<K extends keyof TelemetryMetrics>(key: K, value: number = 1): void {
    if (!this.enabled) return;
    (this.metrics[key] as number) += value;
  }

  async flush(): Promise<void> {
    if (!this.enabled || this.events.length === 0) return;

    const eventsToSend = [...this.events];
    const metricsToSend = { ...this.metrics };
    this.events = [];

    try {
      await invoke('telemetry_flush', {
        events: eventsToSend,
        metrics: metricsToSend,
        sessionId: this.sessionId,
      });
    } catch (error) {
      console.error('Failed to flush telemetry:', error);
      this.events.unshift(...eventsToSend);
    }
  }

  getSessionId(): string {
    return this.sessionId;
  }

  isEnabled(): boolean {
    return this.enabled;
  }

  setEnabled(enabled: boolean): void {
    this.enabled = enabled;
    if (enabled && !this.flushInterval) {
      this.startFlushInterval();
    } else if (!enabled && this.flushInterval) {
      this.stop();
    }
  }
}

export const telemetryService = new TelemetryService();
export default telemetryService;

File: src-tauri/src/commands/telemetry.rs
use serde::{Deserialize, Serialize};
use tauri::command;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TelemetryEvent {
    pub name: String,
    pub properties: Option<serde_json::Value>,
    pub timestamp: u64,
    pub session_id: String,
    pub version: String,
    pub platform: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TelemetryMetrics {
    pub files_opened: u64,
    pub files_saved: u64,
    pub characters_typed: u64,
    pub commands_executed: u64,
    pub terminals_created: u64,
    pub terminal_commands_run: u64,
    pub git_operations: u64,
    pub git_commits: u64,
    pub git_pushes: u64,
    pub git_pulls: u64,
    pub extensions_installed: u64,
    pub extensions_activated: u64,
    pub ai_queries: u64,
    pub ai_completions_accepted: u64,
    pub ai_completions_rejected: u64,
    pub startup_time: u64,
    pub memory_usage: u64,
    pub cpu_usage: u64,
    pub errors: u64,
    pub crashes: u64,
}

#[command]
pub async fn telemetry_flush(
    events: Vec<TelemetryEvent>,
    metrics: TelemetryMetrics,
    session_id: String,
) -> Result<(), String> {
    // Log telemetry data (in production, this would be sent to an analytics service)
    // For now, we just log it locally
    log::info!(
        "Telemetry flush: session={}, events={}, metrics={:?}",
        session_id,
        events.len(),
        metrics
    );

    // In a real implementation, you would:
    // 1. Write to local log file for debugging
    // 2. Optionally send to analytics endpoint if configured
    // 3. Respect user privacy settings

    Ok(())
}